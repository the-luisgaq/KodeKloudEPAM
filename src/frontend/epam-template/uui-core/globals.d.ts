/// <reference path="../uui_globals.d.ts" />
