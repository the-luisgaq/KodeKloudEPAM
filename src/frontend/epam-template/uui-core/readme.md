# UUI-Core package

This package is a part of [EPAM UUI](https://github.com/epam/UUI) library.

`@epam/uui-core` contains a core parts of the UUI library

It contains:
- Typescript interfaces to allow different UUI packages inter-operate
- Services to call API, handle errors, manage modals and notification, etc.
- State management primitives - Forms, validation, Data Sources
- hooks - `useForm`, `useTableState`, `useVirtualList` and others
- helper functions