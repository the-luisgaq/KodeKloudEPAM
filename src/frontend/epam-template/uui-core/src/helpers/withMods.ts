import * as React from 'react';
import { CX, IHasCX } from '../types';
import { forwardRef } from './forwardRef';

export function withMods<TSource, TResult>(
    Component: React.ComponentType<TSource>,
    getCx?: (props: Readonly<TResult>) => CX,
    getProps?: (props: TResult) => Partial<TSource>,
): React.ComponentType<TResult & React.RefAttributes<any>> {
    const isClassComponent = Component.prototype instanceof React.Component;

    // Class components needs to be wrapped in any case,
    // as they might rely on withMods to add forwardedRef.
    if (!isClassComponent && !getProps && !getCx) {
        return Component as any;
    }

    function applyMods(props: Readonly<TResult>) {
        // Most components are wrapped in withMods component.
        // Please keep this method simple and performant.
        // Don't clone objects/arrays if not needed.

        const modProps = getProps?.(props);
        const result: Readonly<TResult> = Object.assign({}, props, modProps);

        const cx = getCx?.(props);

        if (cx) {
            (result as IHasCX).cx = [cx, (props as any).cx];
        }

        return result;
    }

    let wrappedComponent: React.ComponentType<TResult & React.RefAttributes<any>>;

    if (!Component) {
        // Happens in tests, probably due to circular dependencies.
        return null;
    } else if (isClassComponent) {
        // Class component. Wrap with forwardRef, and pass ref in the forwardedRef prop
        wrappedComponent = forwardRef<any, Readonly<TResult>>((props, ref) => {
            const allProps: any = applyMods(props);
            allProps.forwardedRef = ref;
            return React.createElement(Component, allProps);
        });
    } else if ((Component as any).render) {
        // React.forwardRef component.
        // Basically its object like { $$type: Symbol(FORWARD_REF), render: (props, ref) => ... }
        // However, $$type is not exposed, so there's no good way to detect this.
        // We re-wrap it in another forward ref, to avoid unnecessary stacking two forwardRefs
        wrappedComponent = forwardRef(
            (props: Readonly<TResult>, ref: any) => (Component as any).render(applyMods(props), ref),
        );
    } else if (Component instanceof Function) {
        // Plain functional component. Just wrap with function, and apply mods to props
        wrappedComponent = (props: TResult) => (Component as Function)(applyMods(props));
    } else {
        // Any other type of component. E.g. React.memo.
        // Wrap it in another functional component
        wrappedComponent = forwardRef<any, Readonly<TResult>>((props: TResult, ref) => {
            const resultProps = applyMods(props);
            (resultProps as any).ref = ref;
            return React.createElement(Component, resultProps);
        });
    }

    wrappedComponent.displayName = `${Component?.displayName || Component?.name || 'unknown'} (withMods)`;

    return wrappedComponent;
}
