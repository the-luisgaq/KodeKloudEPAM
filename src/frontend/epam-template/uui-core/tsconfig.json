{
  "extends": "../tsconfig.json",
  "strictNullChecks": true,
  "compilerOptions": {
    "rootDir": ".",
    "outDir": "./build"
  },
  "exclude": [
    "**/__tests__",
    "build",
    "node_modules"
  ]
}
