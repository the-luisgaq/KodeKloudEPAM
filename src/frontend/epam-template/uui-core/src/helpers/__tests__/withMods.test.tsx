import React from 'react';
import { render } from '@epam/uui-test-utils';
import { withMods } from '../withMods';
import { IHasCaption, IHasCX } from '@epam/uui-core';

interface BaseProps {
    caption?: string;
    className?: string;
}

interface ModdedProps extends IHasCaption, IHasCX {}

const Base = ({ caption, className }: BaseProps) => <div className={ className }>{caption}</div>;
const Modded = withMods<BaseProps, ModdedProps>(Base, () => ({ cx: 'class-from-props' }));

describe('withMods', () => {
    it('should apply mods correctly', () => {
        const { asFragment } = render(<Modded cx="class-from-props" caption="test" />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should apply mods without additional props', () => {
        const { asFragment } = render(<Modded cx="class-from-props" />);
        expect(asFragment()).toMatchSnapshot();
    });
});
