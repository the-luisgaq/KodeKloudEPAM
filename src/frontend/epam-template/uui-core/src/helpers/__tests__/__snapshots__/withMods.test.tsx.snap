// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withMods should apply mods correctly 1`] = `
<DocumentFragment>
  <div>
    test
  </div>
</DocumentFragment>
`;

exports[`withMods should apply mods without additional props 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
